### PR TITLE
Fix call to thrust::reduce_by_key in argmin/argmax libcudf groupby

### LINKED ISSUE #9156: [BUG] IllegalMemoryAccess sometimes on sorted group by min of strings
**Describe the bug**
So I have been trying to implement some list aggregations using the existing sorted group by cudf code. At least until #9135 is implemented. I see failures when doing this for strings about half the time with IllegalMemoryAccess errors. I also see similar errors doing max on strings too.

**Steps/Code to reproduce bug**

I know it is a little convoluted to make it happen, but I finally got a reproducible case in C++.

You need to apply [patch.txt](https://github.com/rapidsai/cudf/files/7086994/patch.txt) and unzip [data.zip](https://github.com/rapidsai/cudf/files/7086999/data.zip) placing the parquet file in whatever directory you are going to run `gtests/GROUPBY_TEST` from. The patch disables most of the tests, but also adds a test for min that will read in the parquet file and then do an explode and finally a min aggregation.  The exact steps to make this happen are a bit difficult, which is why the test is written the way it is so that it is close to how the java code was releasing things as it went.

About half to a third of the time I see it crash with an error like.
```
[ RUN      ] groupby_min_string_test.min_sorted_after_explode
unknown file: Failure
C++ exception with description "reduce_by_key: failed to synchronize: cudaErrorIllegalAddress: an illegal memory access was encountered" thrown in the test body.
[  FAILED  ] groupby_min_string_test.min_sorted_after_explode (42 ms)
```

With other failures after it because the illegal memory access was not cleared.  When I run with cuda-memcheck I get [log.txt](https://github.com/rapidsai/cudf/files/7087020/log.txt).

Not sure if this is an error in the aggregation code or if something in explode_outer is not working properly. I cannot reproduce it if I save the data after the explode_outer and do the aggregation just from the raw data.

### PR DESCRIPTION
**Pull request by davidwendt**
Closes #9156 

This PR simplifies the parameters when calling `thrust::reduce_by_key` for the argmin/argmax aggregations in cudf::groupby. The `illegalMemoryAccess` found in #9156 was due to invalid data being passed from `thrust::reduce_by_key` through to the BinaryPredicate function as documented in https://github.com/NVIDIA/thrust/issues/1525

The invalid data being passed is only a real issue for strings columns where the device pointer was neither nullptr nor a valid address. The new logic provides only `size_type` values to `thrust::reduce_by_key` so invalid values can only be out-of-bounds for the input column which is easily checked before retrieving the `string_view` objects within the ArgMin and ArgMax operators.

### CODE REVIEW

**Comment by davidwendt** [comment]
@revans2 Could you try this out to see if this fixes the issue for you as reported in #9156?

**Comment by revans2** [comment]
> @revans2 Could you try this out to see if this fixes the issue for you as reported in #9156?

Yes this fixed the issues I was seeing.

**Comment by davidwendt** [comment]
> > @revans2 Could you try this out to see if this fixes the issue for you as reported in #9156?
> 
> Yes this fixed the issues I was seeing.

After discussing in our libcudf standup, we will move this into 21.10.

File: cpp/src/groupby/sort/group_single_pass_reduction_util.cuh [hunk 1]
```diff
@@ -31,77 +31,50 @@
...
-  template <typename T>
-  __device__ size_type operator()(thrust::tuple<T, size_type> const& rhs) const
+  column_device_view const d_col;
+  CUDA_DEVICE_CALLABLE auto operator()(size_type const& lhs, size_type const& rhs) const
   {
-    return thrust::get<1>(rhs);
+    // The extra bounds checking is due to issue github.com/rapidsai/cudf/9156 and
+    // github.com/NVIDIA/thrust/issues/1525
+    // where invalid random values may be passed here by thrust::reduce_by_key
+    if (lhs < 0 || lhs >= d_col.size() || d_col.is_null(lhs)) { return rhs; }
```
Comment (jrhemstad) [inline cpp/src/groupby/sort/group_single_pass_reduction_util.cuh:53]: So the new logic doesn't reference the sentinel values anymore. Is that what we want?

Comment (davidwendt) [reply cpp/src/groupby/sort/group_single_pass_reduction_util.cuh:53]: The sentinel values were set to -1 to only identify the null values. The nulls are accessed directly the logic here in place of the sentinel.

Comment (jrhemstad) [reply cpp/src/groupby/sort/group_single_pass_reduction_util.cuh:53]: Oh, nice. I guess we only really needed the sentinel for the hash implementation.

**Review by jrhemstad — APPROVED** [review]

**Comment by davidwendt** [comment]
rerun tests

**Review by robertmaynard — CHANGES_REQUESTED** [review]
This looks to be branched from 21.12 but targeting 21.10, you will need to rebase on 21.10

**Comment by davidwendt** [comment]
I seemed to have hosed my branch now. I will close the PR and reopen based on 21.10

**Pull request closed without merging**